prober/utils.go

```diff
@@ -60,31 +55,45 @@ func chooseProtocol(IPProtocol string, fallbackIPProtocol bool, target string, r
 		probeDNSLookupTimeSeconds.Add(lookupTime)
 	}()
 
-	ip, err = net.ResolveIPAddr(IPProtocol, target)
+	resolver := &net.Resolver{}
+	ips, err := resolver.LookupIPAddr(ctx, target)
 	if err != nil {
-		if !fallbackIPProtocol {
-			level.Error(logger).Log("msg", "Resolution with IP protocol failed (fallback_ip_protocol is false):", "err", err)
-		} else {
-			level.Warn(logger).Log("msg", "Resolution with IP protocol failed, attempting fallback protocol", "fallback_protocol", fallbackProtocol, "err", err)
-			ip, err = net.ResolveIPAddr(fallbackProtocol, target)
-		}
+		level.Error(logger).Log("msg", "Resolution with IP protocol failed", "err", err)
+		return nil, 0.0, err
+	}
 
-		if err != nil {
-			if IPProtocol == "ip6" {
-				probeIPProtocolGauge.Set(6)
-			} else {
+	// Return the IP in the requested protocol format
+	var fallback net.IPAddr
+	for _, ip := range ips {
+		switch IPProtocol {
+		case "ip4":
+			if ip.IP.To4() != nil {
+				level.Info(logger).Log("msg", "Resolved target address", "ip", ip)
 				probeIPProtocolGauge.Set(4)
+				return &ip, lookupTime, nil
 			}
-			return ip, 0.0, err
+
+			// ip4 as fallback
+			fallback = ip
+
+		case "ip6":
+
+			if ip.IP.To4() == nil {
+				level.Info(logger).Log("msg", "Resolved target address", "ip", ip)
+				probeIPProtocolGauge.Set(6)
+				return &ip, lookupTime, nil
+			}
+
+			// ip6 as fallback
+			fallback = ip
 		}
 	}
 
-	if ip.IP.To4() == nil {
-		probeIPProtocolGauge.Set(6)
-	} else {
+	// Use fallback ip protocol
```

Comment:
Shouldn't you check that fallback has been set?

Reply:
Good call